sounddevice.py

```diff
@@ -2349,6 +2376,125 @@ def __init__(self, channel_selectors):
+            device.  For additional information, see
+            https://app.assembla.com/spaces/portaudio/git/source/master/src/hostapi/coreaudio/notes.txt
+        change_device_parameters : bool, optional
+            If ``True``, allows PortAudio to change things like the
+            device's frame size, which allows for much lower latency,
+            but might disrupt the device if other programs are using it,
+            even when you are just Querying the device.  ``False`` is
+            the default.
+        fail_if_conversion_required : bool, optional
+            In combination with the above flag, ``True`` causes the
+            stream opening to fail, unless the exact sample rates are
+            supported by the device.
+        conversion_quality : string, optional
+            This sets MacCore's sample rate conversion quality.
+            Possible values are: ``'Min'``, ``'Low'``, ``'Medium'``,
+            ``'High'``, or ``'Max'``.  ``'Max'`` is the default.
+
+        Example
+        -------
+        This example assumes a device having 94 input and 94 output
+        channels.  Input is from channels 8 and 9, and output will be
+        sent to channels 2 and 3:
+
+        >>> import sounddevice as sd
+        >>>
+        >>> input_channels = [8, 9]
+        >>> output_channels = [2, 3]
+        >>>
+        >>> # For input devices, the channel_map is simply the list of
+        >>> # 0-based index numbers for each channel:
+        >>> ch_map_in = input_channels
+        >>> mc_in = sd.CoreAudioSettings(channel_map=ch_map_in)
+        >>>
+        >>> # For output devices, the list needs to be "inverted":
+        >>> # Populate the array with -1 for unwanted channels:
+        >>> info = query_devices(device, 'output')
+        >>> max_output_channels = info['max_output_channels']
+        >>> ch_map_out = [-1 for _ in range(max_output_channels)]
+        >>> # For this example device, max_output_channels == 94.
+        >>> # Now overwrite any -1's for channels we'll use for output:
+        >>> for ndx in range(len(output_channels)):
+        >>>     ch_map_out[output_channels[ndx]]=ndx
+        >>> mc_out = sd.CoreAudioSettings(channel_map=ch_map_out)
+        >>>
+        >>> sd.playrec(..., channels=2, extra_settings=(mc_in, mc_out))
+        """
+        # Check inputs:
+        conversion_dict = {
+            'Min':    _lib.paMacCoreConversionQualityMin,
+            'Low':    _lib.paMacCoreConversionQualityLow,
+            'Medium': _lib.paMacCoreConversionQualityMedium,
+            'High':   _lib.paMacCoreConversionQualityHigh,
+            'Max':    _lib.paMacCoreConversionQualityMax,
+        }
+
+        if channel_map is not None:
+            if isinstance(channel_map, int):
+                raise TypeError('channel_map must be a list or tuple')
+            if isinstance(channel_map, (list, tuple)) \
+                    and len(channel_map)==0:
```

Comment:
I agree, we should stay on the safe side and raise an error.

And yes, the comment makes sense, since it isn't obvious at first glance why we would check for `int` specifically.

I still don't like the `list` and `tuple` check. It may not make a practical difference now, it is rather a philosophical question.
And theoretically, the CFFI call might allow more iterables in the future, and then your check will not detect if those other iterables are empty.

I still think that the best way to handle this is to check for `int` before the call to `_ffi.new()` and check for `len(...) == 0` afterwards.

This is getting quite ideological, so feel free to refuse my suggestion.